### ISSUE TITLE
RFC: Logging in a different formal, e.g., Forensic Lucid

### ISSUE THREAD

**Issue by smokhov** [airbnb/streamalert#18]
Hi,

Thanks for your Enigma talk and release of the tool.

I would like to change the logging format to Forensic Lucid in my fork in a modular way for investigation purposes and automated reasoning. What would be a better place to extend StreamAlert in that way?

Subsequently, if the code is sufficiently modular, how do I contribute the logger/formatter to you upstream? Do you have some contributor guidelines?

Thanks.

**Comment by mime-frame** [comment]
Hi @smokhov 

> I would like to change the logging format to Forensic Lucid

When you say logging format, are you referring to the format of the log coming into your rule(s) in the Lambda function? Or are you referring to the format of the log when it's sent to an output such as S3/PagerDuty/Slack?

**Comment by smokhov** [comment]
@mime-frame the one that is sent to the output. The incoming one is an interesting idea I have not considered, but would be a separate issue.

**Comment by mime-frame** [comment]
Thanks for the clarification @smokhov !

We'd welcome a PR that allows StreamAlert to send data to `outputs` in different formats (json, lucid, ...).

Since this is a medium-size change, next best step would be to reply to this issue with a high-level overview of the changes you'd be making and how it would work. Then add the label 'RFC' (request for comments) so we can review. Once that lightweight process is complete, you'd be good to hammer out a PR!

**Comment by smokhov** [comment]
Cool, thanks. I'll write up the overview with the requested info when I get back from the conference in more detail.

**Comment by smokhov** [comment]
Here is the overview:

I need the outputs as well as the logger info in a specific formal of Forensic Lucid. It is a language to encode and reason about forensic knowledge of a case under investigations from multiple evidential sources encoded in the common format.

I need both the outputs and internal logger to be changeable in that format directly instead of having post-processing scripts doing conversion after the fact since when confronted with challenges to forensic evidence that eliminates one extra step of having to prove correctness of the converters.

The evidential `observation`s in Forensic Lucid have a lot of similarities in syntax to JSON.

- I'd assume for now the changes to be localized to `stream_alert_output/main.py`
- Need to refactor `json` and `logger` outputs by abstracting an interaction with them into, say, a `StreamAlertLogger` class that can allow formatting of the messages per chosen output format type; this is where matchers/outputs would be
- refactor `stream_alert_output/main.py` to use the new class `StreamAlertLogger` with the format as a parameter (enumerated, e.g., as JSON, FLUCID, etc.; I suppose people can add later say XML, SQL, CSV, whatever they need); JSON would be the default as you have now
- optionally refactor other modules to use the new log formatter of `StreamAlertLogger`

I don't see an issue with the design example you mention of PagerDuty or Slack or S3 + Forensic Lucid, but I might not be understanding well the issue. Custom receivers or integrations there would have to understand Forensic Lucid or print it in the raw, but this is a worry for me to worry. I'd include also syslog as one of the destinations.

I don't have a permission to add a label in the github label sense, so just type it here:
**RFC**.

**Comment by jacknagz** [comment]
hey @smokhov  - Could you provide some specs/examples on Forensic Lucid format?

> I need both the outputs and internal logger to be changeable in that format directly instead of having post-processing scripts doing conversion after the fact since when confronted with challenges to forensic evidence that eliminates one extra step of having to prove correctness of the converters.

Just to clarify, you want to output the effects of the type conversion (from dict to Forensic Lucid) to logger to verify it was converted correctly? 

> I'd assume for now the changes to be localized to stream_alert_output/main.py

Yes, you just need to create a new function in `StreamOutput` called `_forensic_lucid`.   And then all subsequent rules you write need to have `forensic_lucid` as an output.

> Need to refactor json and logger outputs by abstracting an interaction with them into, say, a StreamAlertLogger class that can allow formatting of the messages per chosen output format type; this is where matchers/outputs would be

`json` and `logger` aren't considered outputs, could you elaborate more?  `logger/logging` is used just for generating logs from the StreamAlert application itself (for debugging, etc).  It's not for outputting to desired incident management/communications services.

> refactor stream_alert_output/main.py to use the new class StreamAlertLogger with the format as a parameter (enumerated, e.g., as JSON, FLUCID, etc.; I suppose people can add later say XML, SQL, CSV, whatever they need); JSON would be the default as you have now

I think for now you might just want to use the method suggested above instead of making a new class for formatting.  The reason is that every API will have specific formats they need, verses generalized JSON, CSV, etc.  Type conversion can be pretty easily implemented per output by using built-in libraries.

**Comment by jacknagz** [comment]
Hey @smokhov, closing this out since we haven't heard from you.   Please let us know if you would like to follow up at a later time.
